### ISSUE TITLE
Issue with snapshots when unique key is varbinary (0.19)

### ISSUE THREAD

**Issue by b-per** [dbt-msft/dbt-sqlserver#108]
Hi!

Just testing the move from dbt-sqlserver=0.18 to dbt-sqlserver=0.19 and some snapshots that are working in v0.18 are failing now. This is the case when the unique_key of the snapshot is already hashed and of the type `varbinary`.

Here is the error: 
```
('42000', '[42000] [Microsoft][ODBC Driver 17 for SQL Server][SQL Server]The style 121 is not supported for conversions from varbinary to varchar. (9809) (SQLExecDirectW)')
```

The workaround I found is to change `unique_key='my_sk'` to `unique_key='CONVERT(NVARCHAR(32),my_sk,2)'` in the snapshot config but I am wondering if the snapshot macro couldn't be updated instead.

**Comment by mikaelene** [comment]
Hi, I can understand the error, but not why you didn't get an error before. If I remember correct, dbt is making a new key called SCD_ID or similar that is a hash of converting the key you supplied "my_sk" with some date columns etc. It can probably be so that this is not supported by SQL Server. But why did it work before? This is the macro and I don't think it has changed: https://github.com/dbt-msft/dbt-sqlserver/blob/020be4a3ffef68c1af9a827bda27f3d0e5628901/dbt/include/sqlserver/macros/materializations/snapshot/strategies.sql#L1-L6

**Comment by camfrout** [comment]
I just checked the `strategies.sql` file that I am using before the upgrade to v0.19 and it still has the "old" `cast` instead of `convert`, hence why I didn't have the error until now. I just checked `pip list` and this is the macro shipped with dbt-sqlserver=0.18.0 from pypi.

```
{% macro sqlserver__snapshot_hash_arguments(args) %}
    CONVERT(VARCHAR(32), HashBytes('MD5', {% for arg in args %}
        coalesce(cast({{ arg }} as varchar ), '') {% if not loop.last %} + '|' + {% endif %}
    {% endfor %}), 2)
{% endmacro %}
```

**Comment by mikaelene** [comment]
Sorry! You are right! That was changed to increase the precision in the timestamp used in the snapshot. The previous code truncated the timestamp to second. This made the CI-tests fail for snapshots. Unlikely to happen in real life. https://github.com/dbt-msft/dbt-sqlserver/issues/61. I'll mark this as a bug. If you have an idea of how to solve this, please let me know.

**Comment by camfrout** [comment]
Not sure if this will fix it, but would it be possible to change `GETDATE()` to `GETDATE() AT TIME ZONE 'UTC'` in the CI test? And then revert the macro to use CAST instead of CONVERT?

Casting the date with the timezone to VARCHAR keeps the milliseconds info.
`SELECT CAST(GETDATE() AT TIME ZONE 'UTC' as VARCHAR) AS date_utc_to_varchar` returns `2021-02-16 22:47:40.857 +00:00` for example.

**Comment by mikaelene** [comment]
That is a very good idea. That was basically what I was trying to achieve. I'll have a look in the upcoming days.

**Comment by mikaelene** [comment]
Relates to #117. @camfrout do you think #118 will solve it?

**Comment by camfrout** [comment]
I have not tried to run the new code but from what I can see it would solve the problem, yes.

**Comment by mikaelene** [comment]
closed by #118
